# Microgrid sizing optimization

Following the first example about [Microgrid performance simulation](Microgrid_Wind-Solar.ipynb)
using [Microgrids.jl](https://github.com/Microgrids-X/Microgrids.jl),
this notebook delves into **optimal microgrid sizing**.
This means finding the optimal size (rated power or energy) of each component by solving an optimization problem where the objective function is composed of Key Performance Indicators (KPIs) related to the economic and technical performance of the project.

Like in the previous example, the sizing is done for a microgrid project with *wind* and *solar* sources, plus a *battery* and a *dispatchable generator*.

<img alt='schematic of a microgrid with a symbolic representation of the notion of sizing' src='./images/microgrid_sizing.png' style='height:18em'>

The main steps of the optimal sizing process are:
1. set up the problem (in particular loading project data like prices and time series)
2. run the optimization
3. analyze the results

## About the optimization formulation and process

### Optimization formulation

In this illustrative sizing example, the optimization objective (criterion) is:
- minimize the Levelized Costs of Electricity `lcoe` (€/kWh)
- while keeping the load shedding rate below some threshold (ex.: shedding rate ≤ 1%)

but other indicators could be used as well: rate of renewables above some threshold...

The above problem is a constrained optimization problem. Since we'll use unconstrained optimization solvers, we will reformulate it as: objective = lcoe + penalty for excess of shedding.

### Optimization process

This is a “blackbox” type of optimization problem: the optimization algorithm calls iteratively the Microgrid simulator with sizings to be tested until convergence (or reaching the maximum allowed number of iterations):

<img alt='schematic of microgrid sizing optimization process: iterations call of the simulator by optimizer' src='./images/MG_optim_diagram.png' style='height:18em'>

The code is meant for using the [NLopt](https://nlopt.readthedocs.io/en/latest/) library of optimization solvers (through its [NLopt.jl](https://github.com/JuliaOpt/NLopt.jl/) wrapper), but it could be adapted to other optimizers with minor changes (see remark below about the modularity of the code).

In [2]:
using Microgrids
using NLopt # optimization solvers
using Printf # pretty print results
using BenchmarkTools # for timing performance, optional

## Load Microgrid project data

Loading parameters and time series for a Microgrid project with *wind* and *solar* sources, plus a *battery* and a *dispatchable generator*. 
Values gathered from the [Microgrid_Wind-Solar.ipynb]((Microgrid_Wind-Solar.ipynb)) notebook.

In [3]:
include("./Microgrid_Wind-Solar_data.jl");

[ Info: Precompiling CSV [336ed68f-0bac-5ca0-87d4-7b16caf5d00b]
[ Info: Precompiling DataFrames [a93c6f00-e57d-5684-b7b6-d8193f3e46c0]


Data definition for Microgrid with wind, solar, storage and generator...


## Setting up the cost function (criterion) to be optimized

The key coding objective is to **encapsulate** the microgrid simulators (`simulate` function of `Microgrids.jl` package) into an objective function that can be called by the optimization algorithm, that is which respects its expected interface (here NLopt).

To increase the modularity which facilites using optimization solvers others that NLopt's we implement the encapsulation by **3 nested functions**:

1. Simulation of Microgrid project described by a sizing vector `x` (vector input) → returns all simulation statistics
2. Extract KPIs of interest to build a multi-objective criterion: here lcoe and shedding rate
3. Combine these KPIs as one mono-objective criterion: here LCOE + penalty if shedding rate > shed_max
   - and match the interface expected by NLopt's optimizers

but if one cares more about compactness, this could be assembled into one single function.

### Wrapper of the Microgrid simulator

accept a sizing vector `x`, then create all `Microgrids.jl` components and call `simulate`

In [4]:
"""Simulate the performance of a Microgrid project of size `x`
with x=[power_rated_gen, energy_rated_sto, power_rated_pv, power_rated_wind]

Returns stats, costs
"""
    function simulate_microgrid(x)
    project = Project(lifetime, discount_rate, timestep, "€")
    # Split decision variables (converted MW → kW):
    power_rated_gen = x[1]*1000
    energy_rated_sto = x[2]*1000
    power_rated_pv = x[3]*1000
    power_rated_wind = x[4]*1000

    # Create components
    gen = DispatchableGenerator(power_rated_gen,
        fuel_intercept, fuel_slope, fuel_price,
        investment_price_gen, om_price_gen, lifetime_gen,
        load_ratio_min,
        replacement_price_ratio, salvage_price_ratio, fuel_unit)
    batt = Battery(energy_rated_sto,
        investment_price_sto, om_price_sto, lifetime_sto, lifetime_cycles,
        charge_rate, discharge_rate, loss_factor_sto, SoC_min, SoC_ini,
        replacement_price_ratio, salvage_price_ratio)
    pv = Photovoltaic(power_rated_pv, irradiance,
        investment_price_pv, om_price_pv,
        lifetime_pv, derating_factor_pv,
        replacement_price_ratio, salvage_price_ratio)
    windgen = WindPower(power_rated_wind, cf_wind,
        investment_price_wind, om_price_wind,
        lifetime_wind,
        replacement_price_ratio, salvage_price_ratio)
    mg = Microgrid(project, Pload, gen, batt, [pv, windgen])

    # Launch simulation:
    traj, stats, costs = simulate(mg)

    return stats, costs
end

simulate_microgrid

Test of the simulator wrapper (on a baseline sizing):

In [5]:
# Baseline sizing: same as in Microgrid_Wind-Solar.ipynb notebook
x_base = [power_rated_gen, energy_rated_sto, power_rated_pv, power_rated_wind]/1000.
# run simulation:
stats, costs = simulate_microgrid(x_base)
x_base, costs.lcoe, costs.npc/1e6

LoadError: UndefVarError: `DispatchableGenerator` not defined

### Build the objective functions (criteria)

- first bi-objective function x ↦ (lcoe, shedding rate)(x)
- then wrapped into a mono objective x ↦ J(x) by using a penalty for the excess of shedding rate
  - and match the interface expected by NLopt's optimizers

In [6]:
"Multi-objective criterion for microgrid performance: lcoe, shedding rate"
function obj_multi(x)
    stats, costs = simulate_microgrid(x)
    # Extract KPIs of interest
    lcoe = costs.lcoe # $/kWh
    shed_rate = stats.shed_rate; # in [0,1]
    return lcoe, shed_rate
end

obj_multi

In [6]:
"""Mono-objective criterion: LCOE + penalty if shedding rate > `shed_max`

with signature adapted to NLopt with `grad` as 2nd argument

load shedding penalty threshold `shed_max` should be in  [0,1[
"""
function obj(x, grad, shed_max, w_shed_max=1e5)
    lcoe, shed_rate = obj_multi(x)
    over_shed = shed_rate - shed_max
    if over_shed > 0.0
        penalty = w_shed_max*over_shed
    else
        penalty = 0.0
    end
    J = lcoe + penalty
end

obj

### Tests the objective functions

Sizing being tested:
- baseline sizing from the simulation notebook: perfect quality of service (QoS) with zero load shedding
- baseline modified with a halved generator sizing: very good QoS with a bit of load shedding → not penalized
- small PV and small wind generators only: low LCOE (i.e. the production-only LCOE of these sources) but but extremely bad QoS → huge penalty

In [7]:
# Test:
shed_max = 0.01 # in [0,1]

x_shed = [power_rated_gen/2, energy_rated_sto, power_rated_pv, power_rated_wind]/1000.
x_pv   = [0. 0. 500.   0.]/1000.
x_wind = [0. 0.   0. 500.]/1000.

println("Base. multi: ", obj_multi(x_base), " mono: ", obj(x_base,[], shed_max))
println("Shed. multi: ", obj_multi(x_shed), " mono: ", obj(x_shed,[], shed_max))
println("PV.   multi: ", obj_multi(x_pv), " mono: ", obj(x_pv,[], shed_max))
println("Wind. multi: ", obj_multi(x_wind), " mono: ", obj(x_wind,[], shed_max))

Base. multi: (0.22924812869928668, 0.0) mono: 0.22924812869928668
Shed. multi: (0.20671000160924613, 0.009602858175478355) mono: 0.20671000160924613
PV.   multi: (0.10149685980966677, 0.923547868561659) mono: 91354.8883530257
Wind. multi: (0.10040264224635914, 0.74395737102815) mono: 73395.83750545724


## Optimization

### Setting up the optimization problem

bounds of the design space and starting point: derived from maximal load power

In [8]:
Pload_max = maximum(Pload)

xmin = [0., 0., 1e-3, 0.] # 1e-3 instead of 0.0, because LCOE is NaN if ther is exactly zero generation
x0 = [1.0, 3.0, 3.0, 2.0] * (Pload_max/1000)
xmax = [1.2, 10.0, 10.0, 5.0] * (Pload_max/1000)

4-element Vector{Float64}:
  2.0484
 17.07
 17.07
  8.535

Optionally ban some choices:

In [9]:
# Solar power forbidden (optional)
#x0[3] = 1e-3
#xmax[3] = 1e-3
# Wind power forbidden (optional)
#x0[4] = 0.
#xmax[4] = 0.

Check cost function on `xmin` and `xmax`

In [10]:
obj_multi(xmin), obj_multi(xmax)

((0.10149685980963595, 0.9998470957371233), (0.8229416738277807, 0.0))

### Wrapper of the optimization process

This is an optional step, but recommended to explore easily the impact of the many parameters taken by optimization algorithms.

See [optimization termination conditions](https://nlopt.readthedocs.io/en/latest/NLopt_Introduction/#termination-conditions) in NLopt documention for the meaning of `xtol_rel`

In [11]:
"""Optimize sizing of microgrid based on the `obj` function

Parameters:
- `x0`: initial sizing (for the algorithms which need them)
- `shed_max`: load shedding penalty threshold (same as in `obj`)
- `algo` could be one of LN_SBPLX, GN_DIRECT, GN_ESCH...
- `maxeval`: maximum allowed number of calls to the objective function,
  that is to the microgrid simulation
- `xtol_rel`: termination condition based on relative change of sizing, see NLopt doc.
- `srand`: random number generation seed (for algorithms which use some stochastic search)

Problem bounds are taken as the global variables `xmin`, `xmax`,
but could be added to the parameters as well.
"""
function optim_mg(x0, shed_max, algo=:LN_SBPLX, maxeval=1000, xtol_rel=1e-4, srand=1)
    nx = length(x0) # number of optim variables
    opt = Opt(algo, nx)
    NLopt.srand(srand)
    
    opt.lower_bounds = xmin
    opt.upper_bounds = xmax
    opt.min_objective = (x, grad) -> obj(x, grad, shed_max)
    opt.xtol_rel = xtol_rel
    opt.maxeval = maxeval
    
    (fopt, xopt, ret) = optimize(opt, x0)
    return xopt, ret, opt.numevals
end

optim_mg

### Run optimization & analyze results

In [12]:
algo = :GN_CRS2_LM # could be one of GN_CRS2_LM, GN_DIRECT, GN_ESCH, LN_SBPLX...
shed_max = 0.01 # in [0,1]
maxeval=1000
xopt, ret, numevals = optim_mg(x0, shed_max, algo, maxeval)

@printf("%s algo: %s after %d iterations. \nx*=", algo, ret, numevals)
println(round.(xopt*1000; digits=1)) # kW
lcoe_opt, shed_rate_opt = obj_multi(xopt)
println("LCOE*: ", lcoe_opt)
println("shed*: ", shed_rate_opt)

GN_CRS2_LM algo: MAXEVAL_REACHED after 1000 iterations. 
x*=[889.0, 2013.0, 1570.9, 1354.0]
LCOE*: 0.19072074165935016
shed*: 0.009961277810994507


optional local "polishing":

In [13]:
algo_polish = :LN_SBPLX
xopt_polish, ret, numevals = optim_mg(xopt, shed_max, algo_polish)

@printf("%s polish: %s after %d iterations. \nx*=", algo_polish, ret, numevals)
println(round.(xopt_polish*1000; digits=1)) # kW
lcoe_opt, shed_rate_opt = obj_multi(xopt_polish)
println("LCOE*: ", lcoe_opt)
println("shed*: ", shed_rate_opt)

LN_SBPLX polish: XTOL_REACHED after 260 iterations. 
x*=[888.5, 1881.3, 1570.9, 1354.0]
LCOE*: 0.1906609830904905
shed*: 0.009999998875359207


Retrieve all performance statistics of the optimized sizing

In [14]:
stats_opt, costs_opt = simulate_microgrid(xopt)
@printf("NPC*: %.2f M\$ (compared to %.2f M\$ in baseline)\n", costs_opt.npc/1e6, costs.npc/1e6)
@printf("rate of renewables: %.1f %%\n", stats_opt.renew_rate*100)
# Display all operation statistics
stats_opt

NPC*: 18.03 M$ (compared to 21.89 M$ in baseline)
rate of renewables: 75.3 %


OperationStats with fields:
- served_energy: 6.7075e6 kWh
- shed_energy: 67487.0 kWh
- shed_max: 680.56 kW
- shed_hours: 370.0 h
- shed_duration_max: 21.0 h
- shed_rate: 0.0099613 in [0,1]
- gen_energy: 1.6592e6 kWh
- gen_hours: 3462.0 h
- gen_fuel: 398200.0 L
- storage_cycles: 190.11 
- storage_char_energy: 401830.0 kWh
- storage_dis_energy: 363560.0 kWh
- storage_loss_energy: 38269.0 kWh
- spilled_energy: 1.2495e6 kWh
- spilled_max: 1941.1 kW
- spilled_rate: 0.1972 in [0,1]
- renew_potential_energy: 6.3361e6 kWh
- renew_energy: 5.0866e6 kWh
- renew_rate: 0.75264 in [0,1]


#### Summary of some results for 1% max shedding rate

Global search algorithms:
- **CRS2**: 0.190 720 \$/kWh (with shedding ~ threshold).
  x* = [889.0, 2013.0, 1570.9, 1354.0] → *best global algorithm choice so far, although effect of seed not studied*
- **DIRECT**: 0.191 297 \$/kWh (with shedding < threshold) in 1000 iter.
  x* = [910.4, 1837.8, 1543.3, 1375.7] 
    - "polishing" with SBPLX: 0.190 650 \$/kWh (with shedding ~ threshold) in 207 extra iter.
      x* = [887.1, 1992.4, 1543.3, 1375.6]
    - increasing DIRECT maxeval to 10k doesn't bring much improvement (0.191 258 \$/kWh)
      compared to SBPLX polishing
- **ESCH**: 0.194 884 \$/kWh with shedding OK (with shedding < threshold) in 1000 iter.
  x* = [1028.1, 1769.6, 1420.5, 1365.1] → *much slower to converge than DIRECT or CRS*

Local search algorithms:
- **SUBPLX**: 0.190 630 \$/kWh with (with shedding ~ threshold) **in 479 iter** (<maxeval).
  x* = [888.6, 1873.5, 1558.5, 1355.9]
  - very competitive, but too sensitive to initialization `x0`?

#### Case with perfect quality of service (QoS)

Another optimization case with perfect quality of service (QoS), that is zero load shedding  

In [15]:
algo = :GN_CRS2_LM # could be one of LN_SBPLX, GN_DIRECT, GN_CRS2_LM, GN_ESCH...
shed_max = 0.00 # in [0,1]
maxeval=1000
xopt, ret, numevals = optim_mg(x0, shed_max, algo, maxeval)

@printf("%s algo: %s after %d iterations. \nx*=", algo, ret, numevals)
println(round.(xopt*1000; digits=1)) # kW
lcoe_opt, shed_rate_opt = obj_multi(xopt)
println("LCOE*: ", lcoe_opt)
println("shed*: ", shed_rate_opt)

GN_CRS2_LM algo: MAXEVAL_REACHED after 1002 iterations. 
x*=[1561.8, 2821.2, 1758.3, 1464.5]
LCOE*: 0.2074272180878332
shed*: 0.0


#### Performance (duration) of the optimization

About 2 s for 10k iterations, i.e. 0.2 ms/iter.

In [16]:
@time xopt, ret, numevals = optim_mg(x0, shed_max, :GN_DIRECT, 10000)

  5.030847 seconds (1.46 M allocations: 7.884 GiB, 10.79% gc time)


([1.5931999999999993, 2.586127114769105, 1.7946595138202435, 1.422494440730462], :MAXEVAL_REACHED, 10000)

Fact: optimization time relates to simulation time with Toptim ~ Tsim × neval

[Benchmarktools](https://juliaci.github.io/BenchmarkTools.jl/)'s `@btime` is need for accurate timing of the simulator:

In [1]:
@btime obj_multi(x0)

LoadError: LoadError: UndefVarError: `@btime` not defined
in expression starting at In[1]:1

## Optimization with Automatic Differentiation *(TO BE WRITTEN)*

As demonstrated in our conference article:

de Godoy Antunes, P. Haessig, C. Wang, and R. Chouhy Leborgne, “Optimal Microgrid Sizing using Gradient-based Algorithms with Automatic Differentiation,” ISGT Europe 2022, Novi Sad, Serbia, 2022. https://dx.doi.org/10.1109/ISGT-Europe54678.2022.9960498 https://hal.archives-ouvertes.fr/hal-03370004

Gradient of optimization criteria computed with [ForwardDiff.jl](https://juliadiff.org/ForwardDiff.jl/stable/).
This allows using **gradient-based optimization algorithms** like LBFGS or SLSQP (the latter supports constraints) which converge in less iterations.

This requires using the "relaxation parameter" `ε` of the Microgrid simulator to smooth out some discontinuities.

DOCUMENTATION TO BE WRITTEN

In [18]:
?aggregation

search: aggregation aggregate



```
aggregation(mg::Microgrid, oper_traj::OperationTraj, ε::Real=1.0)
```

Aggregate operation time series `oper_traj` into yearly statistics for the the microgrid `mg` (returned as an `OperationStats` object).

Discontinuous statistics can optionally be relaxed (smoothed) using the relaxation parameter `ε`:

  * 0.0 means no relaxation (default value)
  * 1.0 yields the strongest relaxation

Using relaxation (`ε` > 0) is recommended when using gradient-based optimization and then a “small enough” value between 0.05 and 0.30 is suggested.


---
## *How we imagine the optimization interface in a future version...*

Desirable feature: *dynamically select* which variables are optimized (e.g. using `getproperty` to dynamically access structure fields...)

In [19]:
#mg0 = baseline microgrid project, with given price and time series data
opti_params = [ # list of :component/variable name, initial, min, max values
    ([:generator,:power_rated], 1000.0, 0., 2000.), 
    ([:battery,:energy_max],    3000.0, 0., 6000.)
]
#optim_mg(mg0, opti_params)

2-element Vector{Tuple{Vector{Symbol}, Float64, Float64, Float64}}:
 ([:generator, :power_rated], 1000.0, 0.0, 2000.0)
 ([:battery, :energy_max], 3000.0, 0.0, 6000.0)